### ISSUE TITLE
Supertraits support

### ISSUE THREAD

**Issue by Lipen** [audunhalland/implementation#1]
I'm currently tinkering with your crate, and my use case is a "service"-like trait with the `Display` bound. Everything boils down to the `impl<T> Display for Impl<T>`, which is impossible to impl because "`Only traits defined in the current crate can be implemented for arbitrary types [E0117]`".

MWE:
```rust
use std::fmt::{Display, Formatter};
use implementation::Impl;

trait MyTrait: Sized + Display { /* service methods */ }

impl<T> MyTrait for Impl<T>
where
    Self: Display, // your crate recommends writing it as `Impl<T>: Display`,
                   // but, as far as I understand, this is equivalent
{}

// Compile error:
//  Only traits defined in the current crate can be implemented for arbitrary types [E0117]
impl<T> Display for Impl<T> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result { todo!() }
}
```

One solution (though, I haven't tested it) is to provide a blanket (or rather, delegating) impl inside `implementation` crate, such as:
```rust
impl<T> Display for Impl<T> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        self.0.fmt(f)  // kinda like that?
    }
}
```
However, this has to be done for all "possible" supertraits, so should not be counted as a "solution".

Another way is to wrap `Impl<T>` with the custom `MyDisplayImpl<T>(Impl<T>)` and use it instead of `Impl<T>: Display`, but it quickly becomes too cumbersome to use, since it does not implement `Deref` etc.

---

More general question is: _does it even make sense_ to mandate `: Display` (or any other supertrait) for `MyTrait` (service-like trait with "single" impl) in the context described in your crate?
My current main problem is not "coherence" (which your crate happens to be solving), but "object-safe trait with generic methods" (which is forbidden) that requires extracting "object-safe" part into a separate service-like trait (thus the `Display` bound, for `dyn MyTrait` to be printable by itself) for which I wanted to use `implementation::Impl` to provide a "unique" impl.

If supertraits are not supported and/or not make sense, it is better to specify it in readme/docs.

**Comment by audunhalland** [comment]
I think you might be right that `Impl` should have `derive`d all well-known traits, including `Debug`, at least. I think `Debug` just slipped my mind when I released version `0.1`. I think I can add that impl and release a ~`0.1.1`~ `0.1.2`.

As for `Display`, I'm actually not sure how to handle that correctly. Should `Display for Impl<T>` always show the same information as `Display for T`? There is an ambiguity issue there, as discribed briefly [here](https://doc.rust-lang.org/rust-by-example/hello/print/print_display.html). I'm not completely sure why you would need `MyTrait` to implement `Display`, maybe you could do the abstraction design in another way, e.g. using your own "print-like" trait instead of `Display`. I'm not sure.

In the case of other traits that can have meaningful blanket impls, but are not as "well known" (3rd-party, i.e. not from the standard library, and not in your own crate) I don't think there is any other solution than to start adding feature flags. Are there any specific traits of this kind that you were thinking of?

**Comment by Lipen** [comment]
> ...I think I can add `Debug` impl...

Yeah, auto-derive `Debug` is probably enough in this case.


> Should Display for Impl<T> always show the same information as Display for T?

I would blindly say "yes", can't see a problem here.


> There is an ambiguity issue there, as discribed briefly [here](https://doc.rust-lang.org/rust-by-example/hello/print/print_display.html).

Hmm, as far I can tell, `Vec<T>` is only similar to `Impl<T>` is a sense they are both _containers_. However, `Vec<T>` is arbitrary-sized container, so the "standard" `Display` implementation for `Vec<T>` is troublesome, to this I agree. On the other hand, `Impl<T>` is just a thin (I would even say `#[repr(transparent)]`) wrapper around `T`. It doesn't add any new functionality. It just delegates to the inner `T`.


> I'm not completely sure why you would need `MyTrait` to implement `Display`, maybe you could do the abstraction design in another way, e.g. using your own "print-like" trait instead of `Display`

That's exactly what I already have (`fn signature() -> &'static str`), and `MyTrait`'s implementors use it inside their `Display` impls. Basically, I am allowed to just use `format!("{}", obj)`, which is not(!) quite equivalent to `format!("{}", obj.signature())`.
To the record, I've also explored an option to add `fn display(&self) -> MyDisplay<'_, T>` (similar to [`std::path::Display`](https://doc.rust-lang.org/std/path/struct.Display.html)). That way I would not need `MyTrait: Display` bound.

Anyway, I ended up not using `Impl` at all, and removed the `MyTrait: Display` bound, replacing it with the on-call bound (`fn do<T: MyTrait+Display>(obj: MyTrait) { ... }`). However, tinkering with `Impl` helper me understand my situation better :)


> In the case of other traits that can have meaningful blanket impls, but are not as "well known" (3rd-party, i.e. not from the standard library, and not in your own crate) I don't think there is any other solution

Basically I want
```rust
impl<Trait, T> Trait for Impl<T> where T: Trait { /* delegate to self.0 */ }
```
which is not possible, I know.


> I'm not sure.
> Are there any specific traits of this kind that you were thinking of?

Me neither :)
Leave it as it is. `Debug` is enough, I think.
